hibernate-core/src/main/java/org/hibernate/bytecode/enhance/internal/bytebuddy/OverridingClassFileLocator.java

```diff
@@ -15,19 +15,34 @@
 /**
  * Allows wrapping another ClassFileLocator to add the ability to define
  * resolution overrides for specific resources.
+ * This is useful when enhancing entities and we need to process an
+ * input byte array representing the bytecode of the entity, and some
+ * external party is providing the byte array explicitly, avoiding for
+ * us having to load it from the classloader.
+ * We'll still need to load several other symbols from a parent @{@link ClassFileLocator}
+ * (typically based on the classloader), but need to return the provided
+ * byte array for some selected resources.
+ * Any override is scoped to the current thread; this is to avoid
+ * interference among threads in systems which perform parallel
+ * class enhancement, for example containers requiring "on the fly"
+ * transformation of classes as they are being loaded will often
+ * perform such operations concurrently.
  */
-public final class OverridingClassFileLocator implements ClassFileLocator {
+final class OverridingClassFileLocator implements ClassFileLocator {
 
-	private final ConcurrentHashMap<String, Resolution> registeredResolutions = new ConcurrentHashMap<>();
+	private final ThreadLocal<ConcurrentHashMap<String, Resolution>> registeredResolutions = ThreadLocal.withInitial( () -> new ConcurrentHashMap<>() );
```

Comment:
`ThreadLocal` and `ConcurrentHashMap`? Can you explain a bit why that makes sense? Also, who is going to cleanup that `ThreadLocal`? It does look like `close` is invoked AFAICS.

Reply:
it's not static, so it will get discarded when we're done with this enhancer

Reply:
and yeah good point about CHM+tlocal being overkill. I'll look at that.